pkg/apis/crd/v1alpha2/types.go

```diff
@@ -389,3 +389,116 @@ type IPPoolList struct {
 
 	Items []IPPool `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// TrafficControl allows mirroring or redirecting the traffic Pods send or receive. It enables users to monitor and
+// analyze Pod traffic, and to enforce custom network protections for Pods with fine-grained control over network
+// traffic.
+type TrafficControl struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard metadata of the object.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of TrafficControl.
+	Spec TrafficControlSpec `json:"spec"`
+}
+
+type TrafficControlSpec struct {
+	// AppliedTo selects Pods to which the traffic control configuration will be applied.
+	AppliedTo AppliedTo `json:"appliedTo"`
+
+	// The direction of traffic that should be matched. It can be Ingress, Egress, or Both.
+	Direction Direction `json:"direction"`
+
+	// The action that should be taken for the traffic. It can be Redirect or Mirror.
+	Action Action `json:"action"`
+
+	// The device to which the traffic should be redirected or mirrored.
+	TargetDevice Device `json:"targetDevice"`
+
+	// The device from which the the traffic will be sent back to OVS. It should only be set for Redirect action.
+	ReturnDevice *Device `json:"returnDevice,omitempty"`
+}
+
+type Direction string
+
+const (
+	DirectionIngress Direction = "Ingress"
+	DirectionEgress  Direction = "Egress"
+	DirectionBoth    Direction = "Both"
+)
+
+type Action string
+
+const (
+	ActionRedirect Action = "Redirect"
+	ActionMirror   Action = "Mirror"
+)
+
+// Device represents a network device that can be used as the target device of traffic mirroring or redirecting, and the
+// return device of traffic redirecting.
+// Multiple TrafficControl objects can use same device, however, the devices with same name must have consistent
+// configuration.
+type Device struct {
+	// The name of the device.
+	Name string `json:"returnDevice"`
+	// Port represents a port that is attached to the OVS bridge. It can be an OVS internal port, a physical NIC, or a
+	// veth device.
+	// +optional
+	Port *PortDevice `json:"port,omitempty"`
+	// VXLAN represents a VXLAN tunnel that is created on the Node.
+	// +optional
+	VXLAN *TunnelDevice `json:"vxlan,omitempty"`
```

Comment:
Should we just add a type to TunnelDevice?

Reply:
ERSPAN is kind of Tunnel, if we add a type to TunnelPort, ERSPAN struct would repeat some tunnel attributes.

Reply:
I feel that is ok. Or if we follow ovs-vsctl, we can put ERSPAN parameters into TunnelPort

Reply:
After more thinking, how about:

OVSPort struct {
    Name
    Type // Internal, Device, GRE, GENEVE, VXLAN, ERSPAN
    Tunnel // Tunnel parameters (remote IP, etc.)
    ERSPAN // ERSPAN parameters
}

Reply:
If I get you correctly, it uses a type field to indicate type and have all type specific parameters in single struct. I feel ovs-vsctl organizes parameters in this way because it needs a consistent db schema for all kinds (all fields are in same level in form of key value). It seems opposite to how K8s API organizes type specific parameters.

And I feel organizing type specific parameters in the type's own struct is more clear and easier to validate and use. For users, it's clear to see which parameters a kind of port has, no need to understand the meaning of irrelevant fields. For developers, it's easier to consume the type specific fields when handling one type. Otherwise we need comments to explain which fields are needed for which types for both users and developers.

Examples of K8s APIs: 
https://github.com/kubernetes/kubernetes/blob/1108bed7631f545d43530aa697175d243b99610b/pkg/apis/core/types.go#L58
https://github.com/kubernetes/kubernetes/blob/1108bed7631f545d43530aa697175d243b99610b/pkg/apis/core/types.go#L2234

Reply:
Sure. Your new version looks good to me too. I think it is just about sharing and encapsulation.

Reply:
The only thing is LocalPort looks strange, as it is just a flag and it is needed only by internal port.